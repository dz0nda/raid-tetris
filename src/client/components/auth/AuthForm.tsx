import React, { FC } from 'react';
import { useForm } from '@mantine/form';
import { Box, Button, Divider, Group, Paper, Stack, TextInput } from '@mantine/core';

export interface AuthFormValues {
  username: string;
}

interface AuthFormProps {
  onValidated: (values: AuthFormValues) => void;
}

const USERNAME_VALIDATION_MESSAGE = 'Username should include at least 6 characters';

const validateUsername = (username: string) => (username.length <= 6 ? USERNAME_VALIDATION_MESSAGE : null);

export const AuthForm: FC<AuthFormProps> = ({ onValidated }) => {
  const form = useForm<AuthFormValues>({
    initialValues: {
      username: '',
    },
    validate: {
      username: validateUsername,
    },
  });

  return (
    <Paper radius="md" p="xl">
      <form
        onSubmit={form.onSubmit((values) => {
          onValidated(values);
        })}
      >
        <Stack>
          <Group grow position="center">
            <Stack px={20}>
              <TextInput
                id="auth-input-username"
                label="Anonymous"
                placeholder="Your username"
                radius="lg"
                {...form.getInputProps('username')}
              />

              <Box>
                <Divider label="Or sign in with a wallet" labelPosition="center" />
                <Group grow mb="md" mt="md">
                  <Button radius="xl" variant="outline" color="black">
                    Beacon
                  </Button>
                </Group>
              </Box>
            </Stack>
          </Group>

          <Group position="right" mt="xl">
            <Button type="submit" radius="xl">
              Login
            </Button>
          </Group>
        </Stack>
      </form>
    </Paper>
  );
};
