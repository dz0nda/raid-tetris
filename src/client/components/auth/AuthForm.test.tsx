import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
// import '@testing-library/jest-dom/extend-expect';
import { AuthForm } from './AuthForm';

it('matches snapshot', () => {
  const tree = renderer.create(<AuthForm onValidated={jest.fn()} />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('<AuthForm />', () => {
  const mockOnValidated = jest.fn();

  beforeEach(() => {
    render(<AuthForm onValidated={mockOnValidated} />);
  });

  it('renders correctly', () => {
    const input = screen.getByPlaceholderText('Your username');
    const loginButton = screen.getByText('Login');
    const beaconButton = screen.getByText('Beacon');

    expect(input).toBeInTheDocument();
    expect(loginButton).toBeInTheDocument();
    expect(beaconButton).toBeInTheDocument();
  });

  it('shows validation error for short username', () => {
    const input = screen.getByPlaceholderText('Your username');
    const loginButton = screen.getByText('Login');

    fireEvent.change(input, { target: { value: 'short' } });
    fireEvent.click(loginButton);

    expect(screen.getByText('Username should include at least 6 characters')).toBeInTheDocument();
  });

  it('calls onValidated with correct values on form submission', () => {
    const input = screen.getByPlaceholderText('Your username');
    const loginButton = screen.getByText('Login');

    fireEvent.change(input, { target: { value: 'validUsername' } });
    fireEvent.click(loginButton);

    expect(mockOnValidated).toHaveBeenCalledWith({ username: 'validUsername' });
  });
});
