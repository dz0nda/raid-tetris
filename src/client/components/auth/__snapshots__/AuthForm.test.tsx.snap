// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="mantine-Paper-root mantine-1kp3ejf"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="mantine-Stack-root mantine-caxjnw"
    >
      <div
        className="mantine-Group-root mantine-1uvkm47"
      >
        <div
          className="mantine-Stack-root mantine-1lp970e"
        >
          <div
            className="mantine-InputWrapper-root mantine-TextInput-root mantine-1ejqehl"
          >
            <label
              className="mantine-InputWrapper-label mantine-TextInput-label mantine-1fzet7j"
              htmlFor="auth-input-username"
              id="auth-input-username-label"
              onMouseDown={[Function]}
            >
              Anonymous
            </label>
            <div
              className="mantine-Input-wrapper mantine-TextInput-wrapper mantine-7c7vou"
            >
              <input
                aria-invalid={false}
                className="mantine-Input-input mantine-TextInput-input mantine-e237sj"
                id="auth-input-username"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Your username"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="mantine-1avyp1d"
          >
            <div
              className="mantine-Divider-root mantine-Divider-horizontal mantine-Divider-withLabel mantine-1xhm9ux"
              role="separator"
            >
              <div
                className="mantine-Text-root mantine-Divider-label mantine-Divider-labelDefaultStyles mantine-n5j5v3"
              >
                Or sign in with a wallet
              </div>
            </div>
            <div
              className="mantine-Group-root mantine-1pg17b9"
            >
              <button
                className="mantine-UnstyledButton-root mantine-Button-root mantine-yddmfv"
                data-button={true}
                type="button"
              >
                <div
                  className="mantine-1wpc1xj mantine-Button-inner"
                >
                  <span
                    className="mantine-1ryt1ht mantine-Button-label"
                    style={
                      {
                        "textTransform": undefined,
                      }
                    }
                  >
                    Beacon
                  </span>
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="mantine-Group-root mantine-w8seur"
      >
        <button
          className="mantine-UnstyledButton-root mantine-Button-root mantine-1xgg2sc"
          data-button={true}
          type="submit"
        >
          <div
            className="mantine-1wpc1xj mantine-Button-inner"
          >
            <span
              className="mantine-1ryt1ht mantine-Button-label"
              style={
                {
                  "textTransform": undefined,
                }
              }
            >
              Login
            </span>
          </div>
        </button>
      </div>
    </div>
  </form>
</div>
`;
